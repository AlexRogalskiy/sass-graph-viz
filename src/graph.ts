import * as path from "path";

export class Graph {
  private graph: { [parent: string]: string[] } = {};
  private orphanNodes = new Set<string>();

  addNode(node: string) {
    if (this.graph[node]) return;

    this.graph[node] = [];
    this.orphanNodes.add(node);
  }

  addEdge(parent: string, child: string) {
    this.addNode(parent);
    this.addNode(child);
    this.graph[parent].push(child);
    this.orphanNodes.delete(child);
  }

  getNodes(): Node[] {
    return Object.keys(this.graph).map((nodeName) => this.createNode(nodeName));
  }

  getEdges(): Node[][] {
    const edges: Node[][] = [];

    Object.keys(this.graph).forEach((parent) => {
      this.graph[parent].forEach((child) => {
        edges.push([this.createNode(parent), this.createNode(child)]);
      });
    });

    return edges;
  }

  focusOnNode(nodeName: string): Graph {
    const graph = new Graph();
    graph.addNode(nodeName);

    this.getEdges().forEach(([parent, child]) => {
      if (
        this.isAncestorOrDescendent(parent.name, nodeName) &&
        this.isAncestorOrDescendent(child.name, nodeName)
      ) {
        graph.addEdge(parent.name, child.name);
      }
    });

    return graph;
  }

  withoutExternals(): Graph {
    const graph = new Graph();

    this.getNodes().forEach((node) => {
      if (!node.isExternal()) graph.addNode(node.name);
    });

    this.getEdges().forEach(([parent, child]) => {
      if (!parent.isExternal() && !child.isExternal()) {
        graph.addEdge(parent.name, child.name);
      }
    });

    return graph;
  }

  toString(): string {
    return JSON.stringify(this.graph, null, 2);
  }

  private createNode(nodeName: string): Node {
    return this.isOrphan(nodeName)
      ? new OrphanNode(nodeName)
      : new Node(nodeName);
  }

  private isOrphan(nodeName: string): boolean {
    return this.orphanNodes.has(nodeName);
  }

  private isAncestorOrDescendent(node: string, target: string): boolean {
    return this.isAncestor(node, target) || this.isDescendent(node, target);
  }

  private isAncestor(node: string, target: string): boolean {
    return this.isDescendent(target, node);
  }

  private isDescendent(
    node: string,
    target: string,
    visitedTargets: string[] = [],
  ): boolean {
    // Don't visit twice a target for performance. Avoid blowing call stack.
    if (visitedTargets.includes(target)) return false;
    visitedTargets.push(target);

    if (node === target) return true;

    const targetChildren = this.graph[target];
    if (!targetChildren) return false;

    return (
      targetChildren.includes(node) ||
      targetChildren.some((child) =>
        this.isDescendent(node, child, visitedTargets),
      )
    );
  }
}

class Node {
  readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  isPartial(): boolean {
    return path.basename(this.name).startsWith("_");
  }

  isExternal(): boolean {
    return this.name.startsWith("..");
  }

  isOrphan(): boolean {
    return false;
  }
}

class OrphanNode extends Node {
  isOrphan() {
    return true;
  }
}
