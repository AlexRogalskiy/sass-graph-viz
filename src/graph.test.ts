import { Graph } from "./graph";

it("should add a node", () => {
  const graph = new Graph();

  graph.addNode("main");

  const nodes = graph.getNodes().map((node) => node.name);
  expect(nodes).toEqual(["main"]);
});

it("should add nodes when adding an edge", () => {
  const graph = new Graph();

  graph.addEdge("main", "_header");

  const nodes = graph.getNodes().map((node) => node.name);
  expect(nodes).toEqual(["main", "_header"]);
});

it("should ignore adding a node that already exists", () => {
  const graph = new Graph();

  graph.addEdge("main", "_header");
  graph.addNode("main");

  const edgesCount = graph.getEdges().length;
  expect(edgesCount).toBe(1);
});

it("should identify orphan nodes", () => {
  const graph = new Graph();
  graph.addEdge("main", "_header");
  graph.addNode("_base");

  const nodes = graph.getNodes();
  const mainNode = nodes.find((node) => node.name === "main");
  const headerNode = nodes.find((node) => node.name === "_header");
  const baseNode = nodes.find((node) => node.name === "_base");

  expect(mainNode.isOrphan()).toBe(true);
  expect(headerNode.isOrphan()).toBe(false);
  expect(baseNode.isOrphan()).toBe(true);
});

it("should return graph edges", () => {
  const graph = new Graph();
  graph.addEdge("main", "_base");
  graph.addEdge("main", "_header");
  graph.addEdge("_header", "_colors");
  graph.addEdge("_colors", "_variables");

  const edges = graph
    .getEdges()
    // Resolve nodes name so we can compare.
    .map(([parent, child]) => [parent.name, child.name]);

  expect(edges).toEqual([
    ["main", "_base"],
    ["main", "_header"],
    ["_header", "_colors"],
    ["_colors", "_variables"],
  ]);
});

describe("focus on a node", () => {
  const nodeName = "_footer";

  it("should filter out nodes that are not linked to the given one", () => {
    const graph = new Graph();
    graph.addEdge("main", "_base");
    graph.addEdge("main", "_header");
    graph.addEdge("_base", nodeName);
    graph.addEdge("_header", "_colors");
    graph.addEdge(nodeName, "_colors");
    graph.addEdge("_colors", "_variables");

    const filteredGraph = graph.focusOnNode(nodeName);

    const expectedGraph = new Graph();
    expectedGraph.addEdge("main", "_base");
    expectedGraph.addEdge("_base", nodeName);
    expectedGraph.addEdge(nodeName, "_colors");
    expectedGraph.addEdge("_colors", "_variables");
    expect(filteredGraph).toEqual(expectedGraph);
  });

  it("should handle cyclic dependencies", () => {
    const graph = new Graph();
    graph.addEdge("main", nodeName);
    graph.addEdge(nodeName, "_colors");
    graph.addEdge("_colors", "_variables");
    graph.addEdge("_variables", "_colors");

    expect(() => graph.focusOnNode(nodeName)).not.toThrow();
  });

  it("should keep an isolated node", () => {
    const graph = new Graph();
    graph.addNode("main");

    const filteredGraph = graph.focusOnNode("main");

    expect(filteredGraph).toEqual(graph);
  });
});

describe("without externals", () => {
  it("should filter out all externals nodes", () => {
    const graph = new Graph();
    graph.addEdge("main", "_header");
    graph.addEdge("main", "../_external-file");
    graph.addEdge("_header", "_colors");
    graph.addEdge("_header", "../../node_modules/bootstrap/main");
    graph.addEdge(
      "../../node_modules/bootstrap/main",
      "../../node_modules/bootstrap/_variables",
    );
    graph.addEdge("../_external_partial", "_variables");

    const filteredGraph = graph.withoutExternals();

    const expectedGraph = new Graph();
    expectedGraph.addEdge("main", "_header");
    expectedGraph.addEdge("_header", "_colors");
    // `_variables` is internal node + child of external node
    // => isolated node in the filtered graph.
    expectedGraph.addNode("_variables");
    expect(filteredGraph).toEqual(expectedGraph);
  });

  it("should not filter out isolated nodes", () => {
    const graph = new Graph();
    graph.addNode("_isolated_partial");
    graph.addEdge("main", "_header");
    graph.addEdge("main", "../_external-file");

    const filteredGraph = graph.withoutExternals();

    const expectedGraph = new Graph();
    expectedGraph.addNode("_isolated_partial");
    expectedGraph.addEdge("main", "_header");
    expect(filteredGraph).toEqual(expectedGraph);
  });
});
