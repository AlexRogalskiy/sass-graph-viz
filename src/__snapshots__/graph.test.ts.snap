// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return a string representation of the graph 1`] = `
"{
  \\"main\\": [
    \\"_base\\",
    \\"_header\\"
  ],
  \\"_base\\": [],
  \\"_header\\": [
    \\"_colors\\"
  ],
  \\"_colors\\": [
    \\"_variables\\"
  ],
  \\"_variables\\": []
}"
`;
